tests/test_class_sh_basic.cpp

```diff
@@ -64,6 +64,28 @@ std::unique_ptr<atyp const, sddc> rtrn_udcp() { return std::unique_ptr<atyp cons
 std::string pass_udmp(std::unique_ptr<atyp,       sddm> obj) { return "pass_udmp:" + obj->mtxt; }
 std::string pass_udcp(std::unique_ptr<atyp const, sddc> obj) { return "pass_udcp:" + obj->mtxt; }
 
+
+struct custom_deleter
+{
+    std::string delete_txt;
+    custom_deleter() = default;
+    explicit custom_deleter(const std::string &delete_txt_) : delete_txt(delete_txt_) {}
```

Comment:
These two constructors were missing originally and [caused a ci failure](https://github.com/pybind/pybind11/actions/runs/6762491686/job/18378533319), the error is pretty cryptic;
```
/__w/pybind11/pybind11/include/pybind11/detail/smart_holder_poc.h:125:52: error: no viable conversion from 'pybind11_tests::class_sh_basic::custom_deleter' to 'std::string' (aka 'basic_string<char>')
    explicit custom_deleter(D &&deleter) : deleter{std::forward<D>(deleter)} {}
                                                   ^~~~~~~~~~~~~~~~~~~~~~~~
```

This addition is mostly based on how other structs in this file looked, can't say I fully understand why these are necessary. The centos pipeline in that run also failed, but all other (more recent?) compilers succeeded without these constructors.

Reply:
What do you think about this?

That gives us more to assert on in the test.

https://godbolt.org/z/cz5vP7cWv

(I recommend having `// clang-format on` for the custom deleter code. Maybe move it up. Not sure.)

Reply:
Sure, done with 13105d574094ffa8ed50499a5eb37f9902128694, lots of move constructors, but that probably doesn't hurt.